Fabric.Authorization.API/Modules/UsersModule.cs

```diff
@@ -57,26 +58,52 @@ public class UsersModule : FabricModule<User>
             return groups;
         }
 
-        private async Task<dynamic> AddGranularPermissions(dynamic param, bool denied)
+        private async Task<dynamic> AddGranularPermissions(dynamic param)
         {
-            var apiModel = this.Bind<GranularPermissionApiModel>();
-
-            if (apiModel.Target != param["subjectId"])
-                return CreateFailureResponse("Target must be the subject id.", HttpStatusCode.BadRequest);
+            var apiModel = this.Bind<GranularPermissionApiModel>();            
 
             foreach (var perm in apiModel.Permissions)
                 await CheckAccess(_clientService, perm.Grain, perm.SecurableItem, AuthorizationManageClientsClaim);
 
             var granularPermissions = apiModel.ToGranularPermissionDomainModel();
 
-            if (denied)
-                granularPermissions.DeniedPermissions = apiModel.Permissions.Select(p => p.ToPermissionDomainModel());
-            else
-                granularPermissions.AdditionalPermissions =
-                    apiModel.Permissions.Select(p => p.ToPermissionDomainModel());
+            granularPermissions.IdentityProvider = param.identityProvider;
+            granularPermissions.Id = param.subjectId;
+
+            granularPermissions.DeniedPermissions = apiModel.Permissions
+                .Where(p => p.PermissionAction == PermissionAction.Deny)
+                .Select(p => p.ToPermissionDomainModel());
+            
+            granularPermissions.AdditionalPermissions = apiModel.Permissions
+                .Where(p => p.PermissionAction == PermissionAction.Allow)
+                .Select(p => p.ToPermissionDomainModel());            
 
             await _permissionService.AddUserGranularPermissions(granularPermissions);
             return HttpStatusCode.NoContent;
+        }      
+
+        private async Task<dynamic> DeleteGranularPermissions(dynamic param)
+        {
+            var apiModel = this.Bind<GranularPermissionApiModel>();
+
+            foreach (var perm in apiModel.Permissions)
+                await CheckAccess(_clientService, perm.Grain, perm.SecurableItem, AuthorizationManageClientsClaim);
+
+            var granularPermissions = apiModel.ToGranularPermissionDomainModel();
+
+            granularPermissions.IdentityProvider = param.identityProvider;
+            granularPermissions.Id = param.subjectId;

```

Comment:
Why is the Id the subjectId and not a concatenation of the subjectId and identityProvider? If we want to get permissions for a user, subjectId and identityProvider will be required, right?

Reply:
Yeah thats a good point, it can be a concatenation of both